sql/core/src/main/scala/org/apache/spark/sql/WindowFunctionDefinition.scala

```diff
@@ -0,0 +1,264 @@
+ *       .row
+ *       .following(1)
+ *       .toColumn.as("avg_value"),
+ *     max("value")
+ *       .over
+ *       .partitionBy("k2")
+ *       .orderBy("k3")
+ *       .between
+ *       .preceding(4)
+ *       .following(3)
+ *       .toColumn.as("max_value"))
+ * }}}
+ *
+ *
+ */
+@Experimental
+class WindowFunctionDefinition protected[sql](
+    column: Column = null,
+    partitionSpec: Seq[Expression] = Nil,
+    orderSpec: Seq[SortOrder] = Nil,
+    frame: WindowFrame = UnspecifiedFrame) {
+
+  private[sql] def newColumn(c: Column): WindowFunctionDefinition = {
+    new WindowFunctionDefinition(c, partitionSpec, orderSpec, frame)
+  }
+
+  /**
+   * Returns a new [[WindowFunctionDefinition]] partitioned by the specified column.
+   * {{{
+   *   // The following 2 are equivalent
+   *   df.over.partitionBy("k1", "k2", ...)
+   *   df.over.partitionBy($"K1", $"k2", ...)
+   * }}}
+   * @group window_funcs
+   */
+  @scala.annotation.varargs
+  def partitionBy(colName: String, colNames: String*): WindowFunctionDefinition = {
+    partitionBy((colName +: colNames).map(Column(_)): _*)
+  }
+
+  /**
+   * Returns a new [[WindowFunctionDefinition]] partitioned by the specified column. For example:
+   * {{{
+   *   df.over.partitionBy($"col1", $"col2")
+   * }}}
+   * @group window_funcs
+   */
+  @scala.annotation.varargs
+  def partitionBy(cols: Column*): WindowFunctionDefinition = {
+    new WindowFunctionDefinition(column, cols.map(_.expr), orderSpec, frame)
+  }
+
+  /**
+   * Returns a new [[WindowFunctionDefinition]] sorted by the specified column within
+   * the partition.
+   * {{{
+   *   // The following 2 are equivalent
+   *   df.over.partitionBy("k1").orderBy("k2", "k3")
+   *   df.over.partitionBy("k1").orderBy($"k2", $"k3")
+   * }}}
+   * @group window_funcs
+   */
+  @scala.annotation.varargs
+  def orderBy(colName: String, colNames: String*): WindowFunctionDefinition = {
+    orderBy((colName +: colNames).map(Column(_)): _*)
+  }
+
+  /**
+   * Returns a new [[WindowFunctionDefinition]] sorted by the specified column within
+   * the partition. For example
+   * {{{
+   *   df.over.partitionBy("k1").orderBy($"k2", $"k3")
+   * }}}
+   * @group window_funcs
+   */
+  def orderBy(cols: Column*): WindowFunctionDefinition = {
+    val sortOrder: Seq[SortOrder] = cols.map { col =>
+      col.expr match {
+        case expr: SortOrder =>
+          expr
+        case expr: Expression =>
+          SortOrder(expr, Ascending)
+      }
+    }
+    new WindowFunctionDefinition(column, partitionSpec, sortOrder, frame)
+  }
+
+  /**
+   * Returns a new ranged [[WindowFunctionDefinition]]. For example:
+   * {{{
+   *   df.over.partitionBy("k1").orderBy($"k2", $"k3").between
+   * }}}
+   * @group window_funcs
+   */
+  def between: WindowFunctionDefinition = {
+    new WindowFunctionDefinition(column, partitionSpec, orderSpec,
+      SpecifiedWindowFrame(RangeFrame, UnboundedPreceding, UnboundedFollowing))
```

Comment:
Also i think this is not right, `between` not indicate the frame type is `RangeFrame`